### DISCUSSION TITLE
Getting Access forbidden message when trying to use filter on graphql queries

### DISCUSSION THREAD

**Discussion by sumchans** [Azure/data-api-builder#1423] [General]
I have been trying out the data api builder for azure cosmos db and so far was successfully able to pull data from Azure Cosmos db. The issue I am having now is, when I use filter on the queries I get "Access forbidden to a field referenced in the filter." I believe this has something to do configuring the permissions on the dab-config.json, but I am not able to get a head start so far on this.

Here is my query -
```
query filteritems {
    items(filter: { userPhoneNumber: { contains: "4160000000" } }) {
        items {
            partitionKey
            userPhoneNumber
            userDisplayName
        }
    }
}
```

Here is the result -
```
{
  "errors": [
    {
      "message": "Access forbidden to a field referenced in the filter.",
      "extensions": {
        "code": "AuthorizationCheckFailed"
      }
    }
  ]
}
```

Also attaching my dab-config json file -
```
{
  "$schema": "https://github.com/Azure/data-api-builder/releases/download/v0.6.13/dab.draft.schema.json",
  "data-source": {
    "database-type": "cosmosdb_nosql",
    "options": {
      "database": "Database",
      "schema": "schema.gql"
    },
    "connection-string": "connection-string;"
  },
  "runtime": {
    "graphql": {
      "allow-introspection": true,
      "enabled": true,
      "path": "/graphql"
    },
    "host": {
      "mode": "development",
      "cors": {
        "origins": [],
        "allow-credentials": false
      },
      "authentication": {
        "provider": "StaticWebApps"
      }
    }
  },
  "entities": {
    "Item": {
      "source": "Items",
      "graphql": true,
      "permissions": [
        {
          "role": "anonymous",
          "actions": [
            "*"
          ],
          "fields": {
            "include": [
              "*"
            ]
          }
        }
      ]
    }
  }
}
```

**Comment by dgcaron** [comment]
i am actually running into this to right now, and it looks like it is related to some update as the code i have worked perfectly fine before. (as in last wednesday).

**Reply by sumchans** [reply]
Oh is it. Has this been created as an issue?

**Reply by seantleonard** [reply]
I'm taking a look at this. It looks like it may be how field.include.* is resolved (or lack thereof) for Cosmos.

**Reply by sumchans** [reply]
Okay thanks @seantleonard

**Reply by sumchans** [reply]
@seantleonard Any idea when I would be able to hear more about this issue? I am currently blocked on my project.

**Comment by dgcaron** [comment]
Are you using the database connection for Azure static webapps or are you hosting it in another way. Can you try running it with the previous version and not the latest?

**Reply by sumchans** [reply]
Right now I am calling the API through Postman. In the future this will connected to a mobile app. How do I run it using the previous version?

**Reply by Aniruddh25** [reply]
hi @sumchans , you can install the previous version by specifying the version number in dotnet tool install: 

`dotnet tool install --global Microsoft.DataApiBuilder --version 0.5.35`

**Reply by sumchans** [reply]
Thanks @Aniruddh25 I just tested out with the previous version and I was able to filter items.

**Reply by Aniruddh25** [reply]
@sumchans, Please do note we still don't have field level AuthZ policy support for CosmosDB tracked by #597.
So, you can continue to use it for anonymous users, but not for authenticated users if your application requires that kind of support.

**Reply by sumchans** [reply]
Got it. Thanks @Aniruddh25

**Comment by sumchans** [comment]
Got it thanks @Anirudh

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by dgcaron** [comment]
i have promoted this discussion to an issue as there seems no way around this for cosmosdb with the latest version #1434 

for people using the data-api for azure static webapps, you can still develop locally by installing the 0.35.5 version 

```
dotnet tool install --global Microsoft.DataApiBuilder --version 0.5.35
```

and then run it from the swa-db-connections
```
dab start --config staticwebapp.database.config.json
```

and then use the data-api-url option of the swa cli
```
swa start ./dist  --data-api-devserver-url http://localhost:5000
```

**Comment by tarazou9** [comment]
Thank you for pointing this out and bringing this to our attention! We just had a PR merged for addressing this issue https://github.com/Azure/data-api-builder/pull/1436. We will have this change included in the next release.

**Reply by JonSilver** [reply]
Thanks @tarazou9. Any idea when this will actually hit a release? It would seem it's a showstopper for many.

Alas, of course, even once there's a new release of data-api-builder it'll take quite some time before it's rolled into a new release of static-web-apps-cli, and finds its way to servers. In the meantime, anyone who wants any kind of filtered queries has to abandon any idea of using the SWA data api, or delay their project for some unspecified period. This problem was caused three weeks ago, fixed and merged two weeks ago, but still not released in its own library, let alone finding its way into an upstream release of SWA itself.

Please guys, make sure unit, integration and regression testing is properly done or you'll just go on breaking thousands of live, production systems over and over again. We've already seen this happen with SWA, with a different subsystem that had long ago left preview.

This is the problem with these monolithic hosting offerings where there's many layers of magic happening... if just one layer of the magic has a bug introduced, the whole thing becomes infected. And when there's no system for emergency fix-and-release of dependent software, it causes lots of pain to paying customers.

@Aniruddh25

**Reply by Aniruddh25** [reply]
Absolutely understand your concerns @JonSilver.
#1448(Hotfix for CosmosDb_NoSql filtering issue on March Release) has been merged in, we will be generating a patch and getting it deployed in this week itself. 
Thank you for your patience!

**Reply by joeljeffery** [reply]
@Aniruddh25 First off, genuinely love the work you and your team is doing. Thanks for the update.

Having said that, I'm also affected by this issue. Breaking the filtering for Cosmos DB connections was pretty catastrophic, especially when it transpired that release was pushed through to live Static Web Apps instances. I'm really grateful there's a patch and release coming. But is there anything in your release practices you can change to make sure that in the event of a catastrophic failure like this, you're able to push out a hotfix release and rapidly get that pushed to SWA? 4 weeks of effectively no data connections to Cosmos DB in Static Web Apps is untenable, especially as you had a fix, merged and ready to go, for a couple of weeks now.

**Reply by dgcaron** [reply]
to be fair to the maintainers, the feature is still in preview for static webapps so in practice it shouldn't break any production / critical workloads. on the other hand, promoting a version with this issue isn't really convenient either, hopefully the patch is applied quickly so everybody can move forward from here.

**Reply by joeljeffery** [reply]
I'm totally on board with it being a preview. But preview shouldn't mean that the service is unavailable for the best part of a month because of a commit. 

Anyone can make a commit that breaks something. It can happen to anyone. This is a complex project. Mistakes and unintended consequences are inevitable and thoroughly understandable. 

When the user base noticed, and raised an issue, the maintainers leapt into action and confirmed, then developed a fix. Within a week it seems. All this is good, and to be expected in a functioning, living development.

However, once the scope of the impact was ascertained, please deploy the fix or rollback as a matter of urgency. We need to know that this scenario (an accidental change that kills core functionality) can be mitigated within a reasonable timeframe when it happens again.

The gap between a fix being developed and it being deployed is too great, preview or no preview. It's not OK that the technology is effectively offline for a month, just because it's in "preview".

Having said that, team please keep up the excellent work.

**Reply by dgcaron** [reply]
i believe a fix has been rolled out to static web apps (at least to my instance it seems)

**Reply by joeljeffery** [reply]
That is tremendous news.

**Reply by Aniruddh25** [reply]
@joeljeffery We absolutely understand the need for urgency, we were still in the process of identifying the right hot patch, apologies for the perceived delay. But rest assured, if such a scenario happens again, it will be mitigated in a reasonable timeframe.

**Reply by joeljeffery** [reply]
Thank you @Aniruddh25, I appreciate the response.

**Reply by tarazou9** [reply]
Hey Folks, sorry for the inconvenience, a couple of things that we have addressed and are addressing are as below.
- New DAB release with a mitigation for this issue has now published, this should allow the query filter works for Cosmos. This patch will includes a check to enforce that the runtime config doesn't contain field permissions. So the recommendation is not to have field permissions in the config until the fix for that is ready.
- For connections through Static Web Apps, we have rolled back to the previous version of DAB engine and it should work as how it was working earlier.
- We are actively working on getting in the changes to fix the field level auth support issues, which will be ready in couple of weeks. 

For example,

```diff
"permissions": [
        {
          "role": "anonymous",
          "actions": [
            "create",
-  {
- "action": "read",
- "fields": {
- "include": [
- "*"
-  ],
- "exclude": []
- }
- },
+ "read",
            "update",
            "delete"
          ]
        }
      ]
```
